web/src/main/java/com/epam/rft/atsy/web/controllers/ApplicationStateController.java

```diff
@@ -83,34 +84,36 @@ public ModelAndView loadPage(@RequestParam Long applicationId,
     ModelAndView modelAndView = new ModelAndView(VIEW_NAME);
     modelAndView.addObject("applicationId", applicationId);
 
+    List<StateHistoryDTO> stateHistory=statesHistoryService.getStateHistoriesByApplicationId(applicationId);
+
     List<StateHistoryViewRepresentation>
         stateHistoryViewRepresentations =
-        converterService
-            .convert(statesHistoryService.getStateHistoriesByApplicationId(applicationId),
-                StateHistoryViewRepresentation.class);
+        converterService.convert(stateHistory,StateHistoryViewRepresentation.class);
 
-    if (clickedState != null && isRepresentationsListFirstStateEquals(stateHistoryViewRepresentations, clickedState)) {
+    if (clickedState != null) {
       StateDTO clickedStateDTO = stateService.getStateDtoByName(clickedState);
 
       Assert.notNull(clickedStateDTO);
-
-      StateHistoryViewRepresentation representation = StateHistoryViewRepresentation.builder()
-          .creationDate(new Date())
-          .stateId(clickedStateDTO.getId())
-          .stateName(clickedStateDTO.getName())
-          .build();
-
-      if (clickedState.equals(NEW_STATE)) {
-        ApplicationDTO applicationDTO = applicationsService.getApplicationDtoById(applicationId);
-        representation.setChannelName(
-            channelService.getChannelDtoById(applicationDTO.getChannelId()).getName());
-        representation.setPositionName(
-            positionService.getPositionDtoById(applicationDTO.getPositionId()).getName());
+      if (stateFlowService.isAvailableFromLastState(stateHistory, clickedState)) {
+        StateHistoryViewRepresentation representation = StateHistoryViewRepresentation.builder()
+            .creationDate(new Date())
+            .stateId(clickedStateDTO.getId())
+            .stateName(clickedStateDTO.getName())
+            .build();
+
+        if (clickedState.equals(NEW_STATE)) {
+          ApplicationDTO applicationDTO = applicationsService.getApplicationDtoById(applicationId);
+          representation.setChannelName(
+              channelService.getChannelDtoById(applicationDTO.getChannelId()).getName());
+          representation.setPositionName(
+              positionService.getPositionDtoById(applicationDTO.getPositionId()).getName());
+        }
+
+        stateHistoryViewRepresentations.add(0, representation);
       }
```

Comment:
Include a redirect in the else branch, which sends the browser to the same url, but without the state parameter.